examples/echo_server/tcp_echo_socket.c

```diff
@@ -0,0 +1,207 @@
+#include <string.h>
+#include <microkit.h>
+
+#include "echo.h"
+#include "lwip/ip.h"
+#include "lwip/pbuf.h"
+#include "lwip/tcp.h"
+
+#include <sddf/util/util.h>
+
+// At most ECHO_QUEUE_SIZE - 1 bytes can be in the queue
+#define ECHO_QUEUE_SIZE (TCP_WND + 1)
+
+struct echo_state {
+    bool in_use;
+    // sending ring buffer
+    size_t tail; // data gets added at tail
+    size_t head; // moved forward for acknowledged data
+    char buf[ECHO_QUEUE_SIZE];
+};
+
+// This previously was a LWIP_MEMPOOL, but turns out that doesn't support sizes
+// greater than ~65536 (due to integer overflow somewhere).
+#define MAX_CONCURRENT 1
+static struct echo_state tcp_state_pool[MAX_CONCURRENT];
+
+static struct echo_state *tcp_state_alloc()
+{
+    for (int i = 0; i < MAX_CONCURRENT; ++i) {
+        if (!tcp_state_pool[i].in_use) {
+            tcp_state_pool[i].in_use = true;
+            return &tcp_state_pool[i];
+        }
+    }
+    return NULL;
+}
+
+static void tcp_state_free(struct echo_state *state)
+{
+    assert(state);
+    assert(state->in_use);
+    state->in_use = false;
+}
+
+static size_t queue_space(struct echo_state *state)
+{
+    return (state->head + ECHO_QUEUE_SIZE - state->tail - 1) % ECHO_QUEUE_SIZE;
+}
+
+static size_t queue_cont_space(struct echo_state *state)
+{
+    if (state->tail >= state->head) {
+        return ECHO_QUEUE_SIZE - state->tail;
+    }
+    return state->head - state->tail - 1;
+}
+
+static err_t tcp_echo_sent(void *arg, struct tcp_pcb *pcb, u16_t len)
+{
+    struct echo_state *state = arg;
+
+    state->head = (state->head + len) % ECHO_QUEUE_SIZE;
+
+    // tcp_recved is only for increasing the TCP window, and isn't required to
+    // ACK incoming packets (that is done automatically on receive).
+    tcp_recved(pcb, len);
+
+    return ERR_OK;
+}
+
+static err_t tcp_echo_recv(void *arg, struct tcp_pcb *pcb, struct pbuf *p, err_t err)
+{
+    struct echo_state *state = arg;
+
+    if (p == NULL) {
+        // closing
+        sddf_printf("tcp_echo[%s:%d]: closing\n",
+                    ipaddr_ntoa(&pcb->remote_ip), pcb->remote_port
+                   );
+
+        // TODO is this a use-after-free?
```

Comment:
Could you please look into whether this is a use after free before this PR is merged?